src/Data/Machine/Type.hs

```diff
@@ -54,81 +63,129 @@ module Data.Machine.Type
 
+import Unsafe.Coerce
+
+newtype TranslateT m u k =
+  TranslateT { runTranslateT :: forall t. u t -> MStep m u k t}
+
+type Translate u k = forall m. Monad m => TranslateT m u k
+
+data Step m u k t where
+  Stop :: Step m u k t
+  Yield :: t -> TranslateT m u k -> Step m u k t
+  Await :: (t' -> MStep m u k t) -> (k t') -> MStep m u k t -> Step m u k t
+
+type MStep m u k t = m (Step m u k t)
+
+type MachineT m k o = TranslateT m (Is o) k
+
+instance (Monad m) => Category (TranslateT m) where
+    id :: TranslateT m u u
+    id = TranslateT $ \req -> return $ Await (\t -> return (Yield t id)) req (return Stop)
+    {-# INLINABLE id #-}
+
+    (.) :: forall u v k. TranslateT m u v -> TranslateT m k u -> TranslateT m k v
+    t2 . t1 = TranslateT $ \req -> runTranslateT t1 req >>= go t2
+      where
+        go :: forall t. TranslateT m u v -> Step m k u t -> MStep m k v t
+        go r2 s1 = case s1 of
+            Stop -> return Stop
+            Yield t r1 -> return $ Yield t $ r2 . r1
+            Await g kg fg -> runTranslateT r2 kg >>= go2 g fg
+        go2 :: (t' -> MStep m k u t) -> MStep m k u t -> Step m u v t' -> MStep m k v t
+        go2 g fg s2 = case s2 of
+            Stop -> go stopped =<< fg
+            Yield t r -> go r =<< g t
+            Await h kh fh -> return $ Await (\t -> go2 g fg =<< h t) kh (go2 g fg =<< fh)
+    {-# INLINABLE (.) #-}
+
 -------------------------------------------------------------------------------
 -- Transduction Machines
 -------------------------------------------------------------------------------
 
 -- | This is the base functor for a 'Machine' or 'MachineT'.
 --
 -- Note: A 'Machine' is usually constructed from 'Plan', so it does not need to be CPS'd.
-data Step k o r
-  = Stop
-  | Yield o r
-  | forall t. Await (t -> r) (k t) r
+-- data Step k o r
+--   = Stop
+--   | Yield o r
+--   | forall t. Await (t -> r) (k t) r
 
-instance Functor (Step k o) where
+instance (Functor m) => Functor (Step m u k) where
   fmap _ Stop = Stop
-  fmap f (Yield o k) = Yield o (f k)
-  fmap f (Await g kg fg) = Await (f . g) kg (f fg)
+  fmap f (Yield o k) = Yield (f o) k
+  fmap f (Await g kg fg) = Await (fmap (fmap f) . g) kg (fmap (fmap f) fg)
 
--- | A 'MachineT' reads from a number of inputs and may yield results before stopping
--- with monadic side-effects.
-newtype MachineT m k o = MachineT { runMachineT :: m (Step k o (MachineT m k o)) }
+-- -- | A 'MachineT' reads from a number of inputs and may yield results before stopping
+-- -- with monadic side-effects.
+-- newtype MachineT m k o = MachineT { runMachineT :: m (Step k o (MachineT m k o)) }
 
 -- | A 'Machine' reads from a number of inputs and may yield results before stopping.
 --
 -- A 'Machine' can be used as a @'MachineT' m@ for any @'Monad' m@.
 type Machine k o = forall m. Monad m => MachineT m k o
 
+machineT :: (Functor m) => MStep m (Is o) k o -> MachineT m k o
+machineT m = unsafeCoerce $ \_ -> m
+{-# INLINE machineT #-}
```

Comment:
Whoa... What's this with `unsafeCoerce`? That *certainly* shouldn't appear without very strong justification for

1. Why it's necessary, and
2. Why it's safe.

When answering question 1, please consider using rewrite rules instead. That is, if you can calculate a proof that `a :~: b` but don't want to do it at runtime, you can use `RULES` to replace the call to the proof with `Refl`.

Reply:
I thought this improved the benchmark performance, but the gain seems small. I've pushed a fix